File: Data Structures & Algorithms/subsets/submission-0.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
       vector<vector<int>> ans;
       int ind = 0;
       vector<int> temp = {};
       int f_size = nums.size()-1;

       rec(nums,ind,ans,temp);
       return ans;

    }

    void rec(vector<int>& nums,int ind,vector<vector<int>> &ans,vector<int> t){   
        if (ind == nums.size()){
            ans.push_back(t);
            return;
        }
        t.push_back(nums[ind]);
        ind++;
        rec(nums,ind,ans,t);
        t.pop_back();
        rec(nums,ind,ans,t);
    }


};

File: Data Structures & Algorithms/permutations/submission-0.go
func permute(nums []int) [][]int {
	if len(nums) == 0{
		return [][]int{}
	}

	temp := []int{}
	truth := make([]bool, len(nums)) // if false, number taken
	ans := [][]int{}
	var rec func([]int,  []bool)

	rec = func(temp []int, truth []bool){
		fmt.Println(temp, truth)
		if len(temp) == len(nums){
			d := make([]int, len(nums))
			copy(d, temp)
			ans = append(ans, d)
			return
		}

		fmt.Println("truth_table", truth)
		for ind, v := range truth{
			if v == false{
				truth[ind] = true
				temp = append(temp, nums[ind])
				rec(temp, truth)

				// undo steps to move state upward in tree
				temp = temp[:len(temp)-1]
				truth[ind] = false
			}
		}
	}

	rec(temp, truth)
	return ans
}
